File: microkernel/api/src/main/java/ink/reactor/kernel/logger/Logger.java
package ink.reactor.kernel.logger;

public interface Logger {
    LoggerFormatter getLoggerFormatter();

    void log(final String message);
    void info(final String message);
    void warn(final String message);
    void error(final String message);
    void error(final String message, final Throwable throwable);

    default void log(final String message, Object... toFormat) {
        log(getLoggerFormatter().format(message, toFormat));
    }
    default void info(final String message, Object... toFormat) {
        info(getLoggerFormatter().format(message, toFormat));
    }
    default void warn(final String message, Object... toFormat) {
        warn(getLoggerFormatter().format(message, toFormat));
    }
    default void error(final String message, Object... toFormat) {
        error(getLoggerFormatter().format(message, toFormat));
    }
    default void error(final String message, final Throwable throwable, Object... toFormat) {
        error(getLoggerFormatter().format(message, toFormat), throwable);
    }
}

File: kernel/api/src/main/kotlin/ink/reactor/kernel/logger/Logger.kt
package ink.reactor.kernel.logger

interface Logger {
    val loggerFormatter: LoggerFormatter

    fun debug(message: String)
    fun log(message: String)
    fun info(message: String)
    fun warn(message: String)
    fun error(message: String)
    fun error(message: String, throwable: Throwable)

    fun debug(message: String, vararg toFormat: Any?) {
        debug(this.loggerFormatter.format(message, *toFormat))
    }
    fun log(message: String, vararg toFormat: Any?) {
        log(this.loggerFormatter.format(message, *toFormat))
    }
    fun info(message: String, vararg toFormat: Any?) {
        info(this.loggerFormatter.format(message, *toFormat))
    }
    fun warn(message: String, vararg toFormat: Any?) {
        warn(this.loggerFormatter.format(message, *toFormat))
    }
    fun error(message: String, vararg toFormat: Any?) {
        error(this.loggerFormatter.format(message, *toFormat))
    }
    fun error(message: String, throwable: Throwable, vararg toFormat: Any?) {
        error(this.loggerFormatter.format(message, *toFormat), throwable)
    }
}
